db/migrate/20200514092553_move_katello_fields_from_hostgroups.katello.rb

```diff
@@ -0,0 +1,45 @@
+class MoveKatelloFieldsFromHostgroups < ActiveRecord::Migration[6.0]
+  def up
+    copy_data_from_hostgroup
+
+    change_table :hostgroups do |table|
+      table.remove(
+        :content_source_id,
+        :kickstart_repository_id,
+        :content_view_id,
+        :lifecycle_environment_id
+      )
+    end
+  end
+
+  def down
+    change_table :hostgroups do |t|
+      t.column :kickstart_repository_id, :integer, :null => true
+      t.column :content_source_id, :integer, :null => true
+      t.column :content_view_id, :integer, :null => true
+      t.column :lifecycle_environment_id, :integer, :null => true
+    end
+  end
+
+  def copy_data_from_hostgroup
+    hg_table = ::Hostgroup.arel_table
+    hostgroups = ::Hostgroup.where(
+           hg_table[:content_source_id].eq(nil)
+      .and(hg_table[:kickstart_repository_id].eq(nil))
+      .and(hg_table[:content_view_id].eq(nil))
+      .and(hg_table[:lifecycle_environment_id].eq(nil))
+      .not)
+    hostgroups.find_each do |hostgroup|
+      content_facet = hostgroup.content_facet || hostgroup.build_content_facet
+      content_source_id, kickstart_repository_id, content_view_id, lifecycle_environment_id = Hostgroup.where(
```

Comment:
This change threw me for a loop so I dug into it (the extra lookup didnt make sense) particularly because it still didn't fix the problem for me.

It turns out that because User.current is nil, the hostgroups query is returning nothing. I adjusted the code to set User.current to the anonymous admin and the migration worked perfectly (using the previous approach of taking the values right off the of hostgroup from find_each).

Here's an example:
https://github.com/Katello/katello/blob/master/db/migrate/20150930183738_migrate_content_hosts.rb#L332-L337

Reply:
Interesting findings...
I think that to get around the user filtering it's easier to use unscoped (indeed we don't need any other scope here):

``` ruby
    hostgroups = ::Hostgroup.unscoped.where(
           hg_table[:content_source_id].eq(nil)
      .and(hg_table[:kickstart_repository_id].eq(nil))
      .and(hg_table[:content_view_id].eq(nil))
      .and(hg_table[:lifecycle_environment_id].eq(nil))
      .not)
```
will commit this change in a minute.

As for the second fetch:
I have `delegates` statements that would prevent us from getting to `Hostgroup` parameters and redirect us to the non-existing facet, that's why I wanted to pluck those values directly.
Changed the query to a more presentable code.

Reply:
Actually I tried `unscoped` in my testing yesterday and it didn't find any hostgroups to migrate.

I've been testing the migration in this fashion:

```
::Hostgroup.find(1).update_attributes(lifecycle_environment_id: 1, content_view_id: 3, content_source_id: 1)
...
bundle exec rake db:migrate:up VERSION=20200514092553
bundle exec rake db:migrate:down VERSION=20200514092553
```

Does running the migration in this fashion work for you with unscoped?

Reply:
`update_attributes` does not work as expected. It calls `<attribute>=` method under the hood, so if you have this PR open, it will try to update an attribute on non-existing facet.
Try printing the hostgroup and facet object, you will see they are empty even after the `update_attribute` statement. To actually test this, I had to migrate down, then call pure SQL:
``` ruby
 ActiveRecord::Base.connection.execute('update hostgroups set lifecycle_environment_id = 1 where id=1')
```
After this, the migration ran smoothly.

Reply:
I was seeding data on the hostgroup with `update_attributes` prior to checking out the PR - sorry, I wasn't clear on that :) The attributes were absolutely set before I did the migration